libraries/lib-audio-io/AudioIO.cpp

```diff
@@ -2095,63 +2106,86 @@ bool AudioIO::ProcessPlaybackSlices(
             std::fill_n(pointers[i], mProcessingBuffers[bufferIndex].size(), .0f);
          }
 
-         if (pScope) {
-            const auto discardable = pScope->Process(channelGroup, &pointers[0],
-               mScratchPointers.data(),
-               // The single dummy output buffer:
-               mScratchPointers[mNumPlaybackChannels],
-               mNumPlaybackChannels, len);
-            for(int i = 0; i < seq->NChannels(); ++i)
-            {
-               auto& buffer = mProcessingBuffers[bufferIndex + i];
-               buffer.erase(buffer.begin(), buffer.begin() + discardable);
-            }
+         const auto discardable = pScope->Process(channelGroup, &pointers[0],
+            mScratchPointers.data(),
+            // The single dummy output buffer:
+            mScratchPointers[mNumPlaybackChannels],
+            mNumPlaybackChannels, len);
+         for(int i = 0; i < seq->NChannels(); ++i)
+         {
+            auto& buffer = mProcessingBuffers[bufferIndex + i];
+            buffer.erase(buffer.begin(), buffer.begin() + discardable);
          }
+
          bufferIndex += seq->NChannels();
       }
    }
 
+   //samples at the beginning of could have been discarded
+   //in the previous step, proceed with the number of samples
+   //equal to the shortest buffer size available
    auto samplesAvailable = std::min_element(
       mProcessingBuffers.begin(),
       mProcessingBuffers.end(),
       [](auto& first, auto& second) { return first.size() < second.size(); }
    )->size();
 
-   for(auto& buffer : mMasterBuffers)
-      std::fill_n(buffer.begin(), samplesAvailable, 0);
+   //introduced latency may be too high...
+   //but we don't discard what already was written
+   if(samplesAvailable == 0)
+      return progress;
+
+   //Prepare master buffers.
+   //These buffers may also contain leftovers from the previous pass
+   //Remember initial offset for each of them
+   const auto masterOffsets = stackAllocate(int, mMasterBuffers.size());
```

Comment:
I have doubts about the correctness of the handling of latency for the master channel.
The `discardable` value returned by a real-time effect is the number of leading samples that can be discarded.
For the track effects, this is indeed what is done:
```cpp
         for(int i = 0; i < seq->NChannels(); ++i)
         {
            auto& buffer = mProcessingBuffers[bufferIndex + i];
            buffer.erase(buffer.begin(), buffer.begin() + discardable);
         }
```
For the master track, however, no leading samples are discarded, and then
```cpp
      unsigned bufferIndex = 0;
      for(auto& buffer : mMasterBuffers)
      {
         mPlaybackBuffers[bufferIndex++]->Put(
            reinterpret_cast<constSamplePtr>(buffer.data()),
            floatSample,
            samplesAvailable,
            0
         );
      }
```
which means trailing samples are discarded instead. It was already that way in the previous commit but I missed it.

If I'm correct, the consequence of this mistake would be that, if the master channel has an effect that adds delay, an artefact due to zeroed audio would be heard at the beginning of playback.

Reply:
I also don't see the necessity of `masterOffsets`. I don't think there can be left-overs from one `ProcessPlaybackSlices` to the next, as it would mean this method would produce more audio samples than asked for, and I don't see it to be the case.

Reply:
The problem does exist indeed. In the attached video, you can hear an audible crackle at the beginning of the playback with a non-zero latency effect (TDR Limiter) added to the stack:

https://github.com/audacity/audacity/assets/70063970/742dc31e-0188-4751-b6c6-833cd5279f0d

Reply:
Thanks @dozzzzer .

Reply:
Doing latencies really right is a known bit of technical debt.  I have had ideas.  I didn't find time to elaborate and share them yet.